ads/google/a4a/utils.js

```diff
@@ -124,117 +128,151 @@ export function isReportingEnabled(ampElement) {
 }
 
 /**
+ * Has side-effect of incrementing ifi counter on window.
+ * @param {!../../../extensions/amp-a4a/0.1/amp-a4a.AmpA4A} a4a
+ * @param {!Array<string>=} opt_experimentIds Any experiments IDs (in addition
+ *     to those specified on the ad element) that should be included in the
+ *     request.
+ * @return {!Object<string,null|number|string>} block level parameters
+ */
+export function googleBlockParameters(a4a, opt_experimentIds) {
+  const adElement = a4a.element;
+  const win = a4a.win;
+  win['ampAdGoogleIfiCounter'] = win['ampAdGoogleIfiCounter'] || 1;
+  const slotRect = a4a.getPageLayoutBox();
+  const iframeDepth = iframeNestingDepth(win);
+  // Detect container types.
+  const containerTypeSet = {};
+  for (let el = adElement.parentElement, counter = 0;
+      el && counter < 20; el = el.parentElement, counter++) {
+    const tagName = el.tagName.toUpperCase();
+    if (ValidAdContainerTypes[tagName]) {
+      containerTypeSet[ValidAdContainerTypes[tagName]] = true;
+    }
+  }
+  const pfx =
+      (containerTypeSet[ValidAdContainerTypes['AMP-FX-FLYING-CARPET']]
+       || containerTypeSet[ValidAdContainerTypes['AMP-STICKY-AD']])
+      ? '1' : '0';
+  let eids = adElement.getAttribute('data-experiment-id');
+  if (opt_experimentIds) {
+    eids = mergeExperimentIds(opt_experimentIds, eids);
+  }
+  const containerTypeArray = Object.keys(containerTypeSet);
+  return {
+    'ifi': win['ampAdGoogleIfiCounter']++,
+    'adf': domFingerprint(adElement),
+    'nhd': iframeDepth,
+    'eid': eids,
+    'adx': slotRect.left,
+    'ady': slotRect.top,
+    'oid': '2',
+    pfx,
+    'rc': a4a.fromResumeCallback ? 1 : null,
+    'act': containerTypeArray.length ? containerTypeArray.join() : null,
+  };
+}
+
+/**
+ * @param {!Window} win
+ * @param {string} type matching typing attribute.
+ * @param {!function(!Element):string} groupFn
+ * @return {!Promise<!Object<string,!Array<!Promise<!../../../src/base-element.BaseElement>>>>}
+ */
+export function groupAmpAdsByType(win, type, groupFn) {
+  return resourcesForDoc(win.document).getMeasuredResources(win,
+      r => r.element.tagName == 'AMP-AD' &&
```

Comment:
nit: can you either add {} around the body of the anonymous function or make it all one line?